recipes/full_finetune_distributed.py

```diff
@@ -314,41 +289,51 @@ def _setup_model(
                 ac_option,
             )
 
-        # Wrap the model with FSDP. This will ensure that the model is sharded
-        # across all available GPUs.
-        model = FSDP(
-            module=model,
-            auto_wrap_policy=utils.get_full_finetune_fsdp_wrap_policy(
-                memory_efficient_fsdp_wrap=memory_efficient_fsdp_wrap,
-                modules_to_wrap={modules.TransformerDecoderLayer},
-            ),
-            cpu_offload=CPUOffload(offload_params=fsdp_cpu_offload),
-            sharding_strategy=torch.distributed.fsdp.ShardingStrategy.FULL_SHARD,
-            device_id=self._device,
-            # this recipe does not currently support mixed precision training
-            mixed_precision=None,
-            # Ensure we broadcast params and buffers from rank 0
-            sync_module_states=True,
-            # Initialize empty modules on all non-zero ranks
-            param_init_fn=(
-                lambda module: module.to_empty(
-                    device=torch.device("cuda"), recurse=False
-                )
-                if not self._is_rank_zero
-                else None
-            ),
-        )
-
-        # Ensure no params and buffers are on meta device
-        utils.validate_no_params_on_meta_device(model)
-
-        # original activation checkpointing (full) - flip the condition above
         if enable_activation_checkpointing and ac_mode is None:
             utils.set_activation_checkpointing(
                 model, auto_wrap_policy={modules.TransformerDecoderLayer}
             )
 
+        fsdp_kwargs = {}
+        if fsdp_cpu_offload:
+            fsdp_kwargs["offload_policy"] = CPUOffloadPolicy()
+
+        # Shard the model with FSDP
+        for m in reversed(list(model.modules())):
+            # TransformerDecoderLayer is wrapped by CheckpointWrapper
+            # when enable_activation_checkpointing
+            if enable_activation_checkpointing:
+                if isinstance(m, CheckpointWrapper):
+                    fully_shard(m, **fsdp_kwargs)
+            # For large vocab size, we can additionally shard
+            # the token embeddings and output projections individually
+            if memory_efficient_fsdp_wrap:
+                if isinstance(m, nn.Embedding):
+                    fully_shard(m, **fsdp_kwargs)
+                if isinstance(m, modules.TransformerDecoder):
+                    fully_shard(m.output, **fsdp_kwargs)
+            else:
+                if isinstance(m, modules.TransformerDecoderLayer):
```

Comment:
n00b question: why else?

Should it be:
```
if isinstance(m, modules.TransformerDecoderLayer):
    fully_shard(m, **fsdp_kwargs)
if memory_efficient_fsdp_wrap:
    if isinstance(m, nn.Embedding):
        fully_shard(m, **fsdp_kwargs)
    if isinstance(m, modules.TransformerDecoder):
        fully_shard(m.output, **fsdp_kwargs)

Reply:
Ah I need to update this, the memory efficient wrapping logic should be disentangled from the AC logic. We want to wrap the transformer decoder layer. If activation checkpointing is enabled this is already wrapped in `CheckpointWrapper` so we check that condition. Otherwise we wrap `TransformerDecoderLayer` directly